datalad/support/annexrepo.py

```diff
@@ -1215,11 +1224,29 @@ def call_annex_items_(self, args, files=None, sep=None):
         ------
         See `_call_annex()` for information on Exceptions.
         """
-        out = self._call_annex(
-            args,
-            files=files,
-            protocol=StdOutErrCapture)['stdout']
-        yield from (out.split(sep) if sep else out.splitlines())
+        class GeneratorStdOutErrCapture(GeneratorMixIn, StdOutErrCapture):
+            def __init__(self):
+                GeneratorMixIn.__init__(self)
+                StdOutErrCapture.__init__(self)
+
+            def pipe_data_received(self, fd, data):
+                if fd == 1:
+                    self.send_result(("stdout", data.decode(self.encoding)))
+                    return
+                super().pipe_data_received(fd, data)
+
+        line_splitter = LineSplitter(separator=sep)
+        for source, content in self._call_annex(
+                                args,
+                                files=files,
+                                protocol=GeneratorStdOutErrCapture):
+
+            if source == "stdout":
+                yield from line_splitter.process(content)
+
+        remaining_content = line_splitter.finish_processing()
+        if remaining_content is not None:
+            yield remaining_content + os.linesep
```

Comment:
Note to self: This entire diff block feels like a pattern that we are likely to see repeated frequently. It probably makes sense to pull it out of here.

Reply:
@christian-monch `call_annex_items_()` and `call_git_items_()` were pretty much identical methods before this change. It would make sense to approach that one in this PR too. I think it would give a better perspective on the comment I gave above re the placement of this pattern.

Moreover, hooking `call_git_items_()` to the generator runner would expose it to (in addition):

- `drop`
- `file_has_content()`
- `is_under_annex()`
- `remove()`
- `get_branch_commits_()`
- `get_special_remotes()`
- `diffstatus()`
    - `status`
    - `diff`
- `get_staged_paths()`
- `get_branch_commits()`
- `get_gitattributes()`
- `for_each_ref()`
    - `clone`
    - `get`
    - `is_with_annex()`
    - `get_(remote_)branches()`
    - `get_tags()`
- `call_git_oneline()`
    - `update`

Taken together, this is a vast chunk of datalad functionality, with -- from my POV -- rather limited additional effort. Via `status` and `diff` a good number of other high-level commands are coming in too. It would give us a very concrete idea on how this performs in practice across a wide range of scenarios. WDYT?

Reply:
> - more-or-less comprehensive tests for a BatchedAnnexCommand-style usage, without actually hooking it into the mainline code.

Maybe I misunderstood your comment, but `BatchedAnnex` is based on `BatchedCommand`, which in turn uses the new runner. So all `BatchedAnnex`-operations that are executed are eventually processed by the new runner.



> Worth keeping an eye on the benchmarks. The current sample has a bunch of near-threshold items (in both directions). None of which seem closely related to the above functionality list

Indeed. Looking at the benchmarks was actually one reason to rework the generator-runner approach and remote the "blocking OS-threads". I am surprised about the 1.10 to 1.20 times and will check that out.


> @christian-monch call_annex_items_() and call_git_items_() were pretty much identical methods before this change. It would make sense to approach that one in this PR too. I think it would give a better perspective on the comment I gave above re the placement of this pattern.

I can certainly do it in this PR. I just want to avoid the situation where we are again piling up so many commits that the PR becomes too monstrous.  ;-)    Also, I am not a datalad power user, therefore I might not detect problems that are not caught by the tests. It might be useful to have more people working with this PR before increasing its size too much. Having said that, I agree that it makes a lot of sense the convert `call_git_items_()`, and I will do that.

Reply:
> Maybe I misunderstood your comment, but BatchedAnnex is based on BatchedCommand, which in turn uses the new runner. So all BatchedAnnex-operations that are executed are eventually processed by the new runner.

No, you did not misunderstand. I misread! Thx for pointing this out. I will update the usage overview...

Reply:
re benchmarks: I see ~20% difference report flukes quite frequently. I just wanted to leave a note that the last report at that time had these stats. Nothing to worry about for now, I think.